components/script/dom/readablestream.rs

```diff
@@ -90,266 +118,196 @@ impl ReadableStream {
                 },
             }
         } else {
-            UnderlyingSource::empty()
+            JsUnderlyingSource::empty()
         };
         // TODO
         Err(Error::NotFound)
     }
 
-    fn new_inherited(
-        external_underlying_source: Option<Rc<ExternalUnderlyingSourceController>>,
-    ) -> ReadableStream {
+    fn new_inherited(controller: Controller) -> ReadableStream {
         ReadableStream {
             reflector_: Reflector::new(),
-            js_stream: Heap::default(),
-            js_reader: Heap::default(),
-            has_reader: Default::default(),
-            external_underlying_source,
+            controller: match controller {
+                Controller::ReadableStreamDefaultController(root) => {
+                    ControllerType::Default(Dom::from_ref(&*root))
+                },
+                Controller::ReadableByteStreamController(root) => {
+                    ControllerType::Byte(Dom::from_ref(&*root))
+                },
+            },
+            stored_error: DomRefCell::new(None),
+            disturbed: Default::default(),
+            reader: MutNullableDom::new(None),
+            state: DomRefCell::new(ReadableStreamState::Readable),
         }
     }
 
-    fn new(
-        global: &GlobalScope,
-        external_underlying_source: Option<Rc<ExternalUnderlyingSourceController>>,
-    ) -> DomRoot<ReadableStream> {
-        reflect_dom_object(
-            Box::new(ReadableStream::new_inherited(external_underlying_source)),
-            global,
-        )
+    fn new(global: &GlobalScope, controller: Controller) -> DomRoot<ReadableStream> {
```

Comment:
Step 3 in https://streams.spec.whatwg.org/#rs-constructor will need to create a `ReadableStream` but controller isn't created yet at that time. The controller is set in https://streams.spec.whatwg.org/#set-up-readable-stream-default-controller.

Reply:
Right, that step is still called from within the constructor, but if necessary we can wrap the controller in a `MutNullableDom`.